sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala

```diff
@@ -606,20 +707,28 @@ class AvroStateEncoder(
     keyStateEncoderSpec match {
       case NoPrefixKeyStateEncoderSpec(_) =>
         encodeUnsafeRowToAvro(row, avroEncoder.keySerializer, keyAvroType, out)
+        val avroRow =
+          encodeUnsafeRowToAvro(row, avroEncoder.keySerializer, keyAvroType, out)
+        // prepend stateSchemaId to the Avro-encoded key portion for NoPrefixKeys
+        encodeWithStateSchemaId(
+          StateSchemaIdRow(stateSchemaInfo.get.keySchemaId, avroRow))
       case PrefixKeyScanStateEncoderSpec(_, _) =>
         encodeUnsafeRowToAvro(row, avroEncoder.keySerializer, prefixKeyAvroType, out)
       case _ => throw unsupportedOperationForKeyStateEncoder("encodeKey")
     }
   }
 
   override def encodeRemainingKey(row: UnsafeRow): Array[Byte] = {
-    keyStateEncoderSpec match {
+    val avroRow = keyStateEncoderSpec match {
       case PrefixKeyScanStateEncoderSpec(_, _) =>
         encodeUnsafeRowToAvro(row, avroEncoder.suffixKeySerializer.get, remainingKeyAvroType, out)
       case RangeKeyScanStateEncoderSpec(_, _) =>
         encodeUnsafeRowToAvro(row, avroEncoder.keySerializer, remainingKeyAvroType, out)
       case _ => throw unsupportedOperationForKeyStateEncoder("encodeRemainingKey")
     }
+    // prepend stateSchemaId to the remaining key portion
+    encodeWithStateSchemaId(
```

Comment:
It looks like we have a case of missing schema ID when the schema of prefix/range scan is the same with key schema. Do I understand correctly, or if we address this somewhere, could you please point it out?

Reply:
Added a test case for this